scripts/binary/binary-cleanup.js

```diff
@@ -0,0 +1,190 @@
+const fs = require('fs-extra')
+const path = require('path')
+const { consolidateDeps } = require('@tooling/v8-snapshot')
+const del = require('del')
+const esbuild = require('esbuild')
+const snapshotMetadata = require('@tooling/v8-snapshot/cache/prod-darwin/snapshot-meta.cache.json')
+const tempDir = require('temp-dir')
+const workingDir = path.join(tempDir, 'binary-cleanup-workdir')
+
+fs.ensureDirSync(workingDir)
+
+async function removeEmptyDirectories (directory) {
+  // lstat does not follow symlinks (in contrast to stat)
+  const fileStats = await fs.lstat(directory)
+
+  if (!fileStats.isDirectory()) {
+    return
+  }
+
+  let fileNames = await fs.readdir(directory)
+
+  if (fileNames.length > 0) {
+    const recursiveRemovalPromises = fileNames.map(
+      (fileName) => removeEmptyDirectories(path.join(directory, fileName)),
+    )
+
+    await Promise.all(recursiveRemovalPromises)
+
+    // re-evaluate fileNames; after deleting subdirectory
+    // we may have parent directory empty now
+    fileNames = await fs.readdir(directory)
+  }
+
+  if (fileNames.length === 0) {
+    await fs.rmdir(directory)
+  }
+}
+
+const getDependencyPathsToKeep = async () => {
+  let entryPoints = new Set([
+    // This is the entry point for the server bundle. It will not have access to the snapshot yet. It needs to be kept in the binary
+    require.resolve('@packages/server/index.js'),
+    // This is a dynamic import that is used to load the snapshot require logic. It will not have access to the snapshot yet. They needs to be kept in the binary
+    require.resolve('@packages/server/hook-require.js'),
+    // These dependencies are started in a new process or thread and will not have access to the snapshot. They need to be kept in the binary
+    require.resolve('@packages/server/lib/plugins/child/require_async_child.js'),
+    require.resolve('@packages/server/lib/plugins/child/register_ts_node.js'),
+    require.resolve('@packages/rewriter/lib/threads/worker.ts'),
+    // These dependencies use the `require.resolve(<dependency>, { paths: [<path>] })` pattern where <path> is a path within the cypress monorepo. These will not be
+    // pulled in by esbuild but still need to be kept in the binary.
+    require.resolve('webpack'),
+    require.resolve('webpack-dev-server', { paths: [path.join(__dirname, '..', '..', 'npm', 'webpack-dev-server')] }),
+    require.resolve('html-webpack-plugin-4', { paths: [path.join(__dirname, '..', '..', 'npm', 'webpack-dev-server')] }),
+    require.resolve('html-webpack-plugin-5', { paths: [path.join(__dirname, '..', '..', 'npm', 'webpack-dev-server')] }),
```

Comment:
we use both 4 and 5?

Reply:
Yes. We try and match the webpack version being used by the consuming project